waspc/src/Wasp/Generator/Monad.hs

```diff
@@ -61,24 +64,44 @@ runGenerator generator =
   let (errorOrResult, finalState) = runIdentity $ runStateT (runExceptT (_runGenerator generator)) initialState
    in (warnings finalState, loggedErrorsOrResult (errorOrResult, errors finalState))
   where
-    initialState = GeneratorState {warnings = [], errors = []}
+    initialState = GeneratorState {warnings = [], errors = [], mangledNames = mempty}
 
     loggedErrorsOrResult (Right result, []) = Right result
     loggedErrorsOrResult (Left _, []) = error "Generator produced error, but had empty log - this should never happen!"
     loggedErrorsOrResult (_, loggedErrors) = Left $ fromList loggedErrors
 
 -- This logs a warning but does not short circuit the computation.
 logGeneratorWarning :: GeneratorWarning -> Generator ()
-logGeneratorWarning w = modify $ \GeneratorState {errors = errors', warnings = warnings'} ->
-  GeneratorState {errors = errors', warnings = w : warnings'}
+logGeneratorWarning w = modify $ \GeneratorState {errors = errors', warnings = warnings', mangledNames = mangledNames'} ->
+  GeneratorState {errors = errors', warnings = w : warnings', mangledNames = mangledNames'}
+
+-- Mangles a name to avoid name clashes. This is useful when generating code that may have user-defined names.
+-- It keeps the mangled names unique by appending a counter to the end of the name.
+mangleName :: String -> Generator String
+mangleName originalName = do
+  mangledNamesMap <- gets mangledNames
+
+  let (mangledName, newMangledNamesMap) = mangleName' originalName mangledNamesMap
+
+  modify $ \s -> s {mangledNames = newMangledNamesMap}
+
+  return mangledName
+  where
+    mangleName' :: String -> Data.Map.Map String Int -> (String, Data.Map.Map String Int)
```

Comment:
Ok there is one thing I thought of only now -> and that is that we might want to minimize the impact of them adding a new user defined name to the codebase. If such user defined name shifts all other names for a +1, then we have a lot of files updating in generated code, while we could have easily avoided that. Is that why you used the approach you used? If so, that is a quite good reason, and it might be worthing mentioning it in the comments somewhere, in case somebody in the future gets same idea like me to simplify the code.